File: src/main/java/com/qlm/zombie/player/InfectionHandler.java
package com.qlm.zombie.player;

import com.qlm.zombie.QLMZombieMod;
import com.qlm.zombie.effect.QLMEffects;
import net.minecraft.network.chat.Component;
import net.minecraft.world.effect.MobEffectInstance;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.entity.monster.Zombie;
import net.minecraft.world.entity.monster.ZombieVillager;
import net.minecraft.world.entity.monster.Drowned;
import net.minecraft.world.entity.monster.Husk;
import net.minecraftforge.event.entity.living.LivingHurtEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;

@Mod.EventBusSubscriber(modid = QLMZombieMod.MOD_ID)
public class InfectionHandler {

    private static final float INFECTION_CHANCE = 0.25f;
    private static final int INFECTION_DURATION = 20 * 60 * 3;
    private static final int INFECTION_DURATION_EASY = 20 * 60 * 2;
    private static final int INFECTION_DURATION_HARD = 20 * 60 * 5;

    @SubscribeEvent
    public static void onPlayerHurtByZombie(LivingHurtEvent event) {
        if (!(event.getEntity() instanceof Player player)) return;

        Entity source = event.getSource().getEntity();
        if (source == null) return;

        if (!isZombieType(source)) return;

        if (player.hasEffect(QLMEffects.INFECTION.get())) return;

        float chance = getInfectionChance(source);
        if (player.getRandom().nextFloat() < chance) {
            int duration = getInfectionDuration(source);
            int amplifier = getInfectionAmplifier(source);

            MobEffectInstance infection = new MobEffectInstance(
                    QLMEffects.INFECTION.get(),
                    duration,
                    amplifier,
                    false,
                    true,
                    true
            );
            player.addEffect(infection);

            if (!player.level().isClientSide) {
                player.displayClientMessage(
                        Component.literal("§c☠ 你被感染了！使用解毒剂(右键)清除感染"),
                        true
                );
            }
        }
    }

    private static boolean isZombieType(Entity entity) {
        return entity instanceof Zombie
                || entity instanceof ZombieVillager
                || entity instanceof Drowned
                || entity instanceof Husk
                || entity instanceof com.qlm.zombie.entity.GiantZombieEntity;
    }

    private static float getInfectionChance(Entity source) {
        if (source instanceof com.qlm.zombie.entity.GiantZombieEntity) {
            return 0.80f;
        }
        if (source instanceof Husk) {
            return 0.50f;
        }
        if (source instanceof Drowned) {
            return 0.30f;
        }
        return INFECTION_CHANCE;
    }

    private static int getInfectionDuration(Entity source) {
        if (source instanceof com.qlm.zombie.entity.GiantZombieEntity) {
            return INFECTION_DURATION_HARD;
        }
        if (source instanceof Husk) {
            return INFECTION_DURATION_HARD;
        }
        if (source instanceof Drowned) {
            return INFECTION_DURATION;
        }
        return INFECTION_DURATION_EASY;
    }

    private static int getInfectionAmplifier(Entity source) {
        if (source instanceof com.qlm.zombie.entity.GiantZombieEntity) {
            return 2;
        }
        if (source instanceof Husk) {
            return 1;
        }
        if (source instanceof Drowned) {
            return 1;
        }
        return 0;
    }
}

File: src/main/kotlin/com/qlm/zombie/player/InfectionHandler.kt
package com.qlm.zombie.player

import com.qlm.zombie.QLMZombieMod
import com.qlm.zombie.config.QLMConfig
import net.minecraft.world.effect.MobEffectInstance
import net.minecraft.world.effect.MobEffects
import net.minecraft.world.entity.player.Player
import net.minecraft.world.entity.monster.Zombie
import net.minecraftforge.event.TickEvent
import net.minecraftforge.event.entity.living.LivingHurtEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod

@Mod.EventBusSubscriber(modid = QLMZombieMod.MOD_ID, bus = Mod.EventBusSubscriber.Bus.FORGE)
object InfectionHandler {

    private const val BASE_INFECTION_CHANCE = 0.15f
    private const val MAX_INFECTION_STACKS = 10
    private const val INFECTION_DURATION_TICKS = 20 * 60 * 3
    private const val INFECTION_DECAY_TICKS = 20 * 60 * 5

    private val infectionStacks = mutableMapOf<String, Int>()
    private val lastInfectionTime = mutableMapOf<String, Long>()

    @JvmStatic
    @SubscribeEvent
    fun onLivingHurt(event: LivingHurtEvent) {
        val entity = event.entity
        val source = event.source

        if (entity.level().isClientSide) return
        if (entity !is Player) return

        val attacker = source.entity
        if (attacker !is Zombie) return

        if (!QLMConfig.enableThirst) return

        val baseChance = BASE_INFECTION_CHANCE
        val currentStacks = infectionStacks[entity.stringUUID] ?: 0
        val chance = baseChance + (currentStacks * 0.03f)

        if (entity.level().random.nextFloat() < chance) {
            applyInfection(entity, currentStacks, chance)
        }
    }

    private fun applyInfection(player: Player, currentStacks: Int, chance: Float) {
        val newStacks = (currentStacks + 1).coerceAtMost(MAX_INFECTION_STACKS)
        val uuid = player.stringUUID
        infectionStacks[uuid] = newStacks
        lastInfectionTime[uuid] = player.level().gameTime

        val amplifier = (newStacks - 1).coerceAtLeast(0)
        player.addEffect(
            MobEffectInstance(
                MobEffects.POISON,
                INFECTION_DURATION_TICKS,
                amplifier,
                false,
                true,
                true
            )
        )

        if (newStacks >= 5) {
            player.addEffect(
                MobEffectInstance(
                    MobEffects.WEAKNESS,
                    INFECTION_DURATION_TICKS,
                    amplifier / 2,
                    false,
                    true,
                    true
                )
            )
        }

        if (newStacks >= MAX_INFECTION_STACKS) {
            player.addEffect(
                MobEffectInstance(
                    MobEffects.CONFUSION,
                    INFECTION_DURATION_TICKS / 2,
                    0,
                    false,
                    true,
                    true
                )
            )
        }

        QLMZombieMod.LOGGER.info(
            "[Infection] 玩家 {} 感染! 层数: {}, 总概率: {:.1f}%",
            player.name.string,
            newStacks,
            (chance * 100)
        )
    }

    @JvmStatic
    @SubscribeEvent
    fun onLevelTick(event: TickEvent.PlayerTickEvent) {
        if (event.phase != TickEvent.Phase.END) return
        val player = event.player
        if (player.level().isClientSide) return

        val uuid = player.stringUUID
        val currentStacks = infectionStacks[uuid] ?: return

        val gameTime = player.level().gameTime
        val lastTime = lastInfectionTime[uuid] ?: 0L

        if (gameTime - lastTime > INFECTION_DECAY_TICKS) {
            infectionStacks[uuid] = (currentStacks - 1).coerceAtLeast(0)
            lastInfectionTime[uuid] = gameTime

            if (infectionStacks[uuid] == 0) {
                infectionStacks.remove(uuid)
                lastInfectionTime.remove(uuid)
                player.removeEffect(MobEffects.POISON)
                player.removeEffect(MobEffects.WEAKNESS)
                player.removeEffect(MobEffects.CONFUSION)
                QLMZombieMod.LOGGER.debug("[Infection] 玩家 {} 感染已清除", player.name.string)
            }
        }
    }

    fun getInfectionStacks(player: Player): Int {
        return infectionStacks[player.stringUUID] ?: 0
    }

    fun cureInfection(player: Player) {
        val uuid = player.stringUUID
        infectionStacks.remove(uuid)
        lastInfectionTime.remove(uuid)
        player.removeEffect(MobEffects.POISON)
        player.removeEffect(MobEffects.WEAKNESS)
        player.removeEffect(MobEffects.CONFUSION)
    }

    fun hasInfection(player: Player): Boolean {
        return (infectionStacks[player.stringUUID] ?: 0) > 0
    }
}